plugins/@grouparoo/pardot/src/lib/client.ts

```diff
@@ -0,0 +1,215 @@
+
+    await this.conn.login(username, fullPassword);
+
+    this.request = axios.create({
+      baseURL: `${this.pardotUrl}/api`,
+      headers: {
+        Authorization: `Bearer ${this.conn.accessToken}`,
+        "Pardot-Business-Unit-Id": this.businessUnitId,
+      },
+    });
+  }
+
+  async getCustomFields(): Promise<PardotCustomField[]> {
+    const res = await this.request.get("/customField/version/4/do/query", {
+      params: {
+        format: "json",
+        output: "mobile", // mobile output does not require pagination
+      },
+    });
+
+    const { result } = res.data;
+
+    const mapResult = (field: any): PardotCustomField => ({
+      id: field.id,
+      name: field.name,
+      fieldId: field.field_id,
+      type: field.type,
+      isRecordMultipleResponses: field.is_record_multiple_responses,
+    });
+
+    if (result.total_results === 0) {
+      return [];
+    } else if (result.total_results === 1) {
+      return [mapResult(result.customField)];
+    }
+
+    return result.customField.map(mapResult);
+  }
+
+  async getProspectByEmail(email: string) {
+    try {
+      const res = await this.request.get(
+        `/prospect/version/4/do/read/email/${email}`,
+        {
+          params: {
+            format: "json",
+          },
+        }
+      );
+
+      if (Array.isArray(res.data.prospect)) {
+        return res.data.prospect[0];
+      }
+
+      return res.data.prospect;
+    } catch (error) {
+      if (
+        error.response?.status === 400 &&
+        error.response?.data["@attributes"]?.err_code === 4
+      ) {
+        return null;
+      }
+
+      throw error;
+    }
+  }
+
+  async getProspectById(id: number) {
+    try {
+      const res = await this.request.get(
+        `/prospect/version/4/do/read/id/${id}`,
+        {
+          params: {
+            format: "json",
+          },
+        }
+      );
+
+      if (Array.isArray(res.data.prospect)) {
+        return res.data.prospect[0];
+      }
+
+      return res.data.prospect;
+    } catch (error) {
+      if (
+        error.response?.status === 400 &&
+        error.response?.data["@attributes"]?.err_code === 4
+      ) {
+        return null;
+      }
+
+      throw error;
+    }
+  }
+
+  async queryProspects(
+    options?: Record<string, any>
+  ): Promise<Record<string, any>[]> {
+    const res = await this.request.get("/prospect/version/4/do/query", {
+      params: {
+        format: "json",
+        ...options,
+      },
+    });
+
+    const { result } = res.data;
+    if (result.total_results === 0) {
+      return [];
+    } else if (result.total_results === 1) {
+      return [result.prospect];
+    }
+
+    return result.prospect;
+  }
+
+  async batchUpsertProspects(prospects: Record<string, any>[]) {
+    const res = await this.request.post(
+      "/prospect/version/4/do/batchUpsert",
+      {},
+      {
+        params: {
+          format: "json",
+          prospects: JSON.stringify({
```

Comment:
does`stringify` add any extra spaces? I don't think so but it might be worth checking because the length here can cause problems.

Reply:
Nope, it doesn't